File: myFinance_improved_V2/lib/features/cash_ending/data/repositories/journal_repository_impl.dart
import '../../domain/repositories/journal_repository.dart';
import '../datasources/journal_remote_datasource.dart';

/// Implementation of JournalRepository
/// Coordinates between domain layer and data sources
class JournalRepositoryImpl implements JournalRepository {
  final JournalRemoteDataSource _remoteDataSource;

  JournalRepositoryImpl({
    required JournalRemoteDataSource remoteDataSource,
  }) : _remoteDataSource = remoteDataSource;

  @override
  Future<Map<String, dynamic>> insertJournalWithEverything({
    required double baseAmount,
    required String companyId,
    required String createdBy,
    required String description,
    required String entryDate,
    required List<Map<String, dynamic>> lines,
    String? counterpartyId,
    String? ifCashLocationId,
    String? storeId,
  }) async {
    try {
      return await _remoteDataSource.insertJournalWithEverything(
        baseAmount: baseAmount,
        companyId: companyId,
        createdBy: createdBy,
        description: description,
        entryDate: entryDate,
        lines: lines,
        counterpartyId: counterpartyId,
        ifCashLocationId: ifCashLocationId,
        storeId: storeId,
      );
    } catch (e) {
      // Re-throw with additional context
      throw Exception('Repository error: Failed to insert journal entry - $e');
    }
  }
}

File: website/src/features/cash-ending/data/repositories/JournalRepositoryImpl.ts
/**
 * JournalRepositoryImpl
 * Repository implementation for journal entry operations
 */

import { cashEndingDataSource, ACCOUNT_IDS } from '../datasources/CashEndingDataSource';
import type {
  IJournalRepository,
  InsertJournalParams,
  AccountIds,
} from '../../domain/repositories/IJournalRepository';

export class JournalRepositoryImpl implements IJournalRepository {
  /**
   * Account IDs for journal entries
   */
  readonly accountIds: AccountIds = {
    cash: ACCOUNT_IDS.cash,
    errorAdjustment: ACCOUNT_IDS.errorAdjustment,
    foreignCurrencyTranslation: ACCOUNT_IDS.foreignCurrencyTranslation,
  };

  /**
   * Insert journal entry with all details
   */
  async insertJournalWithEverything(params: InsertJournalParams): Promise<{ success: boolean; error?: string }> {
    try {
      await cashEndingDataSource.insertJournal({
        baseAmount: params.baseAmount,
        companyId: params.companyId,
        createdBy: params.createdBy,
        description: params.description,
        entryDateUtc: params.entryDateUtc,
        lines: params.lines,
        storeId: params.storeId,
      });

      return { success: true };
    } catch (error) {
      const errMsg = error instanceof Error ? error.message : 'Unknown error';
      return { success: false, error: errMsg };
    }
  }
}

export const journalRepository = new JournalRepositoryImpl();
